autoPyTorch/pipeline/components/training/trainer/base_trainer.py

```diff
@@ -0,0 +1,351 @@
+        string += f"\n\t\tEpoch {last_epoch}\n"
+        string += '=' * 40
+        string += "\n"
+        for key, value in sorted(self.performance_tracker.items()):
+            if isinstance(value[last_epoch], dict):
+                # Several metrics can be passed
+                string += "\t{}:\n".format(
+                    key,
+                )
+                for sub_key, sub_value in sorted(value[last_epoch].items()):
+                    string += "\t\t{}: {}\n".format(
+                        sub_key,
+                        sub_value,
+                    )
+            else:
+                string += "\t{}: {}\n".format(
+                    key,
+                    value[last_epoch],
+                )
+        string += '=' * 40
+        return string
+
+
+class BaseTrainerComponent(autoPyTorchTrainingComponent):
+
+    def prepare(
+        self,
+        metrics: List[Any],
+        model: torch.nn.Module,
+        criterion: torch.nn.Module,
+        budget_tracker: BudgetTracker,
+        optimizer: torch.nn.Module,
+        device: torch.device,
+        logger: logging.Logger,
+        writer: Optional[SummaryWriter],
+    ) -> None:
+
+        self.logger = logger
+
+        # Save the device to be used
+        self.device = device
+
+        # Setup the metrics
+        self.metrics = metrics
+
+        # Setup the loss function
+        self.criterion = criterion.to(device)
+
+        # setup the model
+        self.model = model.to(device)
+
+        # setup the optimizers
+        self.optimizer = optimizer
+
+        # The budget tracker
+        self.budget_tracker = budget_tracker
+
+        # A summary writer for tensorboard
+        self.writer = writer
+
+    def on_epoch_start(self, X: Dict[str, Any], epoch: int) -> None:
+        """
+        Optional place holder for AutoPytorch Extensions.
+
+        An user can define what happens on every epoch start or every epoch end.
+        """
+        pass
+
+    def on_epoch_end(self, X: Dict[str, Any], epoch: int) -> bool:
+        """
+        Optional place holder for AutoPytorch Extensions.
+        An user can define what happens on every epoch start or every epoch end.
+        If returns True, the training is stopped
+
+        """
+        return False
+
+    def train(self, train_loader: torch.utils.data.DataLoader, epoch: int
+              ) -> Tuple[float, Dict[str, float]]:
+        '''
+            Trains the model for a single epoch.
+
+        Args:
+            train_loader (torch.utils.data.DataLoader): generator of features/label
+            epoch (int): The current epoch used solely for tracking purposes
+
+        Returns:
+            float: training loss
+            Dict[str, float]: scores for each desired metric
+        '''
+
+        loss_sum = 0.0
+        N = 0
+        self.model.train()
+        outputs_data = list()
+        targets_data = list()
+
+        for step, (data, targets) in enumerate(train_loader):
```

Comment:
This could be outsourced to an additional method `make_step()` such that we could train for a certain number of steps instead of epochs

Reply:
thanks for the suggestion I have created this function